internal/cloudplugin/versions.go

```diff
@@ -0,0 +1,241 @@
+package cloudplugin
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/hashicorp/go-getter"
+	svchost "github.com/hashicorp/terraform-svchost"
+	"github.com/hashicorp/terraform/internal/releaseauth"
+)
+
+// VersionManager downloads, caches, and returns information about versions
+// of terraform-cloudplugin binaries downloaded from the specified backend.
+type VersionManager struct {
+	signingKey         string
+	binaryName         string
+	cloudPluginDataDir string
+	host               svchost.Hostname
+	client             *CloudPluginClient
+	goos               string
+	arch               string
+	ctx                context.Context
+}
+
+// Version is a struct containing the path to the binary corresponding to
+// the manifest version.
+type Version struct {
+	BinaryLocation    string
+	ProductVersion    string
+	ResolvedFromCache bool
+}
+
+const (
+	KB = 1000
+	MB = 1000 * KB
+)
+
+// NewVersionManager initializes a new VersionManager to broker data between the
+// specified directory location containing cloudplugin package data and a
+// Terraform Cloud backend URL.
+func NewVersionManager(ctx context.Context, cloudPluginDataDir string, serviceURL *url.URL, goos, arch string) (*VersionManager, error) {
+	client, err := NewCloudPluginClient(ctx, serviceURL)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize cloudplugin version manager: %w", err)
+	}
+
+	return &VersionManager{
+		cloudPluginDataDir: cloudPluginDataDir,
+		host:               svchost.Hostname(serviceURL.Host),
+		client:             client,
+		binaryName:         "terraform-cloudplugin",
+		goos:               goos,
+		arch:               arch,
+		ctx:                ctx,
+	}, nil
+}
+
+func (v VersionManager) versionedPackageLocation(version string) string {
+	return path.Join(v.cloudPluginDataDir, "bin", version, fmt.Sprintf("%s_%s", v.goos, v.arch))
```

Comment:
For providers we use these versioned URLs to help us detect whether a plugin cache directory is consistent with the dependency lock file and with the configuration, and because we share a common directory structure both for the cache directory in `.terraform/providers` and for local mirror directories where the ability to have multiple versions available is a requirement.

Are there some similar requirements for us to encode the version number of this plugin in the directory structure, and/or to support having multiple versions of the plugin installed at once? I suppose I'd imagined this particular case as being a little simpler where at any given time there is only one plugin version that everyone ought to be using, and they might lag a bit in getting upgraded but there's never any reason to run an older version once a newer version has been installed.

I'm not opposed to this structure, but I'm just wondering if there are some opportunities to simplify this a little and e.g. rely entirely on checksum matching to determine if the plugin ought to be updated, and overwrite the old plugin in-place when we upgrade it. That might make the code in here somewhat simpler -- perhaps we could remove the concept of "version" entirely, for example -- but I dunno, maybe it's not actually gonna make much difference in practice :man_shrugging:

Reply:
You're right about the basic idea here that multiple versions are never needed and this approach may unnecessarily increase storage over time for terraform cloud users. I'd like to rethink this!

Reply:
@apparentlymart I added https://github.com/hashicorp/terraform/pull/33602/commits/8f4a430e418b0ac70c9b90ef05cd8dde1b10bd5c which removes the concept of versions, but still stores a ".version" file in the data directory alongside the binary to compare against.